askama_derive/src/input.rs

```diff
@@ -88,61 +69,76 @@ impl<'a> TemplateInput<'a> {
         match (&source, ext.is_some()) {
             (&Source::Path(_), true) => {
                 panic!("'ext' attribute cannot be used with 'path' attribute")
-            }
+            },
             (&Source::Source(_), false) => {
                 panic!("must include 'ext' attribute when using 'source' attribute")
-            }
-            _ => {}
+            },
+            _ => {},
         }
 
-        let escaping = match escaping {
-            Some(m) => m,
-            None => {
-                let ext = match source {
-                    Source::Path(ref p) => Path::new(p)
-                        .extension()
-                        .map(|s| s.to_str().unwrap())
-                        .unwrap_or(""),
-                    Source::Source(_) => ext.as_ref().unwrap(), // Already panicked if None
-                };
-                if HTML_EXTENSIONS.contains(&ext) {
-                    EscapeMode::Html
-                } else {
-                    EscapeMode::None
-                }
-            }
-        };
-
-        let parent = match ast.data {
-            syn::Data::Struct(syn::DataStruct {
-                fields: syn::Fields::Named(ref fields),
-                ..
-            }) => fields.named.iter().filter_map(|f| {
-                f.ident
-                    .as_ref()
-                    .and_then(|name| if name == "_parent" { Some(&f.ty) } else { None })
-            }),
-            _ => panic!("derive(Template) only works for struct items"),
-        }.next();
-
         let path = match source {
-            Source::Source(_) => match ext {
-                Some(ref v) => PathBuf::from(format!("{}.{}", ast.ident, v)),
-                None => PathBuf::new(),
-            },
-            Source::Path(ref s) => path::find_template_from_path(s, None),
+            Source::Source(_) => PathBuf::from(format!("{}.{}", ast.ident, ext.unwrap())),
+            Source::Path(ref path) => path::find_template_from_path(path, None),
         };
 
         TemplateInput {
             ast,
             source,
             print,
-            escaping,
-            ext,
-            parent,
+            escaping: escaping.unwrap_or(EscapeMode::from(&path)),
+            parent: field_type(ast, "_parent"),
             path,
         }
     }
+
+    pub fn source(&self) -> String {
```

Comment:
I feel like the caller shouldn't care about how it gets the source so this implementation detail got moved here. In a separate function to avoid another clone.

Reply:
I don't like this part much. I like that the `TemplateInput` is just a well-typed representation of the input AST, that represents all of the `template()` attribute stuff as well as anything the code generator might need from the AST. Then `build_template()` can be responsible for taking that input and working out what needs to be done with it, including getting the source from the `path` if necessary. I'd also prefer to keep that in `lib.rs` for the symmetry with included templates, for which a symmetric operation happens in `find_used_templates()`.